File: apps/web/open2log/lib/open2log/products/price.ex
defmodule Open2log.Products.Price do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id
  schema "prices" do
    belongs_to :product, Open2log.Products.Product
    belongs_to :shop, Open2log.Stores.Shop
    belongs_to :user, Open2log.Accounts.User

    # Price info
    field :price_cents, :integer
    field :currency, :string, default: "EUR"

    # Unit price for comparison
    field :unit_price_cents, :integer
    field :comparison_unit, Ecto.Enum, values: [:kg, :l, :pcs]

    # Source tracking
    field :source, Ecto.Enum, values: [:crawled, :user_scanned]
    field :scanned_at, :utc_datetime

    # Images stored in R2
    field :barcode_image_url, :string
    field :price_image_url, :string

    # Validity
    field :valid_from, :utc_datetime
    field :valid_until, :utc_datetime

    timestamps(type: :utc_datetime)
  end

  def changeset(price, attrs) do
    price
    |> cast(attrs, [:product_id, :shop_id, :user_id, :price_cents, :currency,
                    :unit_price_cents, :comparison_unit, :source, :scanned_at,
                    :barcode_image_url, :price_image_url, :valid_from, :valid_until])
    |> validate_required([:product_id, :shop_id, :price_cents, :source])
    |> foreign_key_constraint(:product_id)
    |> foreign_key_constraint(:shop_id)
  end
end

File: apps/ios/Open2Log/Sources/Models/Price.swift
import Foundation

struct Price: Codable, Identifiable {
    let id: UUID
    let productId: UUID
    let shopId: UUID
    let userId: UUID?
    let priceCents: Int
    let currency: String
    let unitPriceCents: Int?
    let comparisonUnit: UnitType?
    let source: PriceSource
    let scannedAt: Date?
    let barcodeImageUrl: String?
    let priceImageUrl: String?
    let validFrom: Date?
    let validUntil: Date?
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case productId = "product_id"
        case shopId = "shop_id"
        case userId = "user_id"
        case priceCents = "price_cents"
        case currency
        case unitPriceCents = "unit_price_cents"
        case comparisonUnit = "comparison_unit"
        case source
        case scannedAt = "scanned_at"
        case barcodeImageUrl = "barcode_image_url"
        case priceImageUrl = "price_image_url"
        case validFrom = "valid_from"
        case validUntil = "valid_until"
        case createdAt = "created_at"
    }

    var priceFormatted: String {
        let euros = Double(priceCents) / 100.0
        return String(format: "%.2f €", euros)
    }
}

enum PriceSource: String, Codable {
    case crawled
    case userScanned = "user_scanned"
}
